### DISCUSSION TITLE
Lazy Load Props for Hydration

### DISCUSSION THREAD

**Discussion by gregod** [Elderjs/elderjs#194] [Ideas]
At the moment, the data required for hydration is written directly as a script tag to the (otherwise static) site. From a no-js first perspective, this prop data seems wasteful as we need to wait for lazily importing the component anyway. Was this done for performance reasons?

I'm proposing to write the prop data to a js module instead of the HTML and then import both data and the component in parallel in the hydration code triggered by the interaction observer. This should perform almost as well while minimizing the js required in the static HTML file.

Maybe an option would be to inline up until a specific size and switch to importing otherwise?

**Comment by nickreese** [comment]
@gregod Good question. The main reason for Elder.js writing it to the HTML has been some users are using Elder.js at the edge which makes deploys easier. 

In `next` which will be `v1.5` (pr here: https://github.com/Elderjs/elderjs/pull/191) I'm experimenting with "prop" compression. Look at the props on this page: https://findenergy.com/tx/. It compresses the props about 45% on large pages before brotli and 10% after brotli. This almost seems silly when I could just be using an external module.

On data heavy pages, I 100% agree that external files are a better answer. I just haven't had time to explore it.

Open to a PR or continuing this discussion.

**Reply by nickreese** [reply]
With more of our sites moving to Elder.js using SSR... one of the issues of using an external file is needing to keep track of the props for a specific page in memory and intercept the requests. 

This can be alleviated by writing the files to the file system... but then we'd probably want to be careful to clean up those files at some point as we'd obviously be hashing the files so they play well with CDNs.

Totally doable with Elder.js' internals, just want to make sure we're considering all of the edge cases.

**Reply by nickreese** [reply]
Stream of consciousness almost... but I think on of the best courses of action is to just assume in both SSR/SSG that only the proper hashed prop files would be served. It is almost outside the scope of Elder.js to manage the prop files. We'll wipe the `/_elderjs/props/` folder on bootstrap and be done with it. This is the same approach we take to hashed css files. 

On SSR, something like `express` or `polka` will serve the file. In SSG only the hashed file will assume to be uploaded.

**Reply by kevmodrome** [reply]
Have you epxlored the idea of putting the props in json files instead of js files (unsure if I mentioned this during the podcast)? JSON files are generally much faster to parse than JS files. At least it was a couple of years ago: https://v8.dev/blog/cost-of-javascript-2019#json

**Reply by gregod** [reply]
Yes, I was also thinking of handling the file similar to the CSS stack, just keyed per hydration.
I'm not too sure where the best place to perform the extraction into the separate would be in the stack.  I have a quick demo that does the writing directly within the hydrate component file (where it probably shouldn't be) that reuses the random prop name as an identifier.  
This works but has different file names on every build.  A proper hash would make this much more CDN friendly. (And could share cache across pages with the same hydration data)

The rough POC:
Instead of pushing the data into the hydration stack with a variable of uniquePropsName,

https://github.com/Elderjs/elderjs/blob/58e027ab8f851992873130b638bea76c31ef1130/src/partialHydration/hydrateComponent.ts#L65-L69

I just write the contents into a js file next to the component file, but keyed by the unique propname. It will simply export the data given to the prop.
```javascript
fs.writeFileSync(`public/${clientSrcMjs}.${uniquePropsName}.js`,`export default ${devalue_1.default(props)};`)
```
Later when hydrating,
https://github.com/Elderjs/elderjs/blob/58e027ab8f851992873130b638bea76c31ef1130/src/partialHydration/hydrateComponent.ts#L103-L108

we can fetch the props data file in parallel and render based on the imported props
```javascript
          Promise.all([import("${clientSrcMjs}"),import("${clientSrcMjs}.${uniquePropsName}.js")]).then(([component,props])=>{
            new component.default({ 
              target: document.getElementById('${uniqueComponentName}'),
              props: ${hasProps ? 'props.default' : '{}'},
              hydrate: true
              });
          });
```


This way, we have one additional file per component instance rendered. This should be manageable, especially if the filenames are relatively consistent.

**Reply by nickreese** [reply]
Nice. I’ve got a similar PoC working with my local 1.5 branch which uses requestIdleCallback and intersectionObserver. 

Props are loaded on requestIdleCallback. Biggest challenges are around ie11 support (which I’m ready to drop) and some thoughts around defaults for each prop option… and if this should be available via prop options, sidewide default, or over a specific size of props. 

Should have time to hack on this tomorrow. Watch this space.

**Reply by gregod** [reply]
Would using json via ajax instead of importing as @kevmodrome  suggested remove any issues with ie? (However, i guess the "devalue" conflicts). Otherwise, maybe a fallback "jsonp style" nomodule js file could be generated, which just injects the uniquePropsName variable? 

I have only used elder.js for a small toy project, so take the following thoughts for defaults (or any other for that matter) with a grain of salt.

For an empty cache scenario, having the prop file external should always be better. The component js is externalized anyway, and if we request both in parallel, props files that are relatively small should complete before the component anyway. Larger prop files are good candidates for externalization anyway. If the component is cached this is a different story.
This feels like the problem faced in inlining critical css. So maybe some ideas from there can be grabbed?

Going even further from the original idea: What if the component gets control over the data fetching? There is probably only a small part of the data that is required immediately on load, while some is only used after user interaction. 
If we do a ```<Component hydrate-client-critical={{}} hydrate-client-a={{}} hydrate-client-b={{}} />```, and inline the critical props while the props of a and b (then not quite props anymore) are only available via a ```` await helpers.load("a")``` within the component?

Edit:
Maybe the last thought could use "raw" routes to simplify the architecture? Instead of having to deal with a special props externalization and data loading code we let the existing props as is but only use it for critical props. Then a new api/raw route type is added which instead of rendering svelte code just runs JS and writes the result to the file. A component could just load the fixed url of the route with a ajax request. 

Come to think of it, a raw route type might be a good idea anyway to generate images etc.

**Reply by nickreese** [reply]
@gregod Interesting thought. I'd argue that components have full control of data fetching with `onMount`. That is how we use it daily.

That said, I did some major surgery to how we hydrate components which will cause a ton of test rewrites... 😢 but I really like the outcome. The only non-spaghetti solution included dropping IE11 support which I should have done a long time ago.

I've integrated both @kevmodrome's `JSON.parse()` suggestion and your suggestion. 

On `elderjs@elderjs@next` you have access a `props.hydrate` key in your `elder.config.js` which is explained on the PR: https://github.com/Elderjs/elderjs/pull/191

This release will be a monster and may take some time as I'll want to get all the tests working again.

**Reply by gregod** [reply]
Very nice!  

Fyi, i had to fix two typos to get it running: [Promise.all ](https://github.com/Elderjs/elderjs/blob/475223e10ae572ae284c829370c7aa659b806bbe/src/partialHydration/hydrateComponent.ts#L16)requires an array of promises and the [noop prop decode](https://github.com/Elderjs/elderjs/blob/475223e10ae572ae284c829370c7aa659b806bbe/src/partialHydration/hydrateComponents.ts#L121) has the wrong variable name. 

I see quite a bit (>50%) of code size optimization potential in the generated loader code (especially the IntersectionObserver). Should i wait a bit until you finalize your major surgery, or would you prefer a proposal/discussion via pull request on the current 1.5 branch?

**Reply by nickreese** [reply]
@gregod Made those two changes and updated `next`. So hard to write js in string literals. 

Major surgery is over. Happy to have a discussion or a PR against v1.5. 👍

**Reply by nickreese** [reply]
@gregod Updating a bunch of tests today. Should make any changes easier to integrate.

**Reply by nickreese** [reply]
Pseudo code I'm working towards with the help of @lukeed 

```html
<script type="module">

  function $$ejs_init(elem, component, props){
    import(component).then((comp)=>{
      new comp.default({ 
        target: elem,
        props: $ejs(props),
        hydrate: true
      });
    });
}

const IO, ITEMS={};
async function $$ejs(id, component, props={}) {
  const elem = document.getElementById(id);
  let propsToHydrate = props;
  if(typeof props === 'string' && props.startsWith('http')){
    const propsFile = await import(props);
    propsToHydrate = propsFile.default;
  }
  
  if (!IO) {
    IO = new IntersectionObserver((entries, observer) => {
      // ...
      if (entry.isIntersecting) {
        observer.unobserve(elem);
        ITEMS[id] = true;
        $$ejs_init(elem, component, props);
      }
    });
  }
  IO.observe(elem);
}

requestIdleCallback(function(){
    $$ejs('headerGppxgYmtDt','/dev/_elderjs/svelte/components/Header/Header.AOWJN766.js',JSON.parse(`{"permalink":"/dev/power-plants/","server":"u"}`));
})

</script>
```

**Reply by nickreese** [reply]
Implemented on `v1.5` and pushed on `next`.

**Reply by nickreese** [reply]
Here is the implementation if curious:

```js
let IO,
  $$COMPONENTS = {};
const $$ejs = async (arr) => {
  $ejs = function (_ejs) {
    return _ejs;
  };
  for (let i = 0; i < arr.length; i++) {
    $$COMPONENTS[arr[i][0]] = {
      elem: document.getElementById(arr[i][0]),
      component: arr[i][1],
      props: arr[i][2] || {},
    };

    if (typeof $$COMPONENTS[arr[i][0]].props === 'string') {
      const propsFile = await import($$COMPONENTS[arr[i][0]].props);
      $$COMPONENTS[arr[i][0]].props = propsFile.default;
    }

    if (!IO) {
      IO = new IntersectionObserver((entries, observer) => {
        var objK = Object.keys(entries);
        var objKl = objK.length;
        var objKi = 0;
        for (; objKi < objKl; objKi++) {
          const entry = entries[objK[objKi]];
          if (entry.isIntersecting) {
            const selected = $$COMPONENTS[entry.target.id];
            observer.unobserve(selected.elem);
            import(selected.component).then((comp) => {
              new comp.default({
                target: selected.elem,
                props: $ejs(selected.props),
                hydrate: true,
              });
            });
          }
        }
      });
    }
    IO.observe($$COMPONENTS[arr[i][0]].elem);
  }
};

requestIdleCallback(
  function () {
    $$ejs([
      [
        'zoomablemappMxdlyXvRt',
        '/dev/_elderjs/svelte/components/ZoomableMap/ZoomableMap.AOMHQNYN.js',
        '/dev/_elderjs/props/ejs-614237837.js',
      ],
      [
        'sortabletableFmqxLmRrQM',
        '/dev/_elderjs/svelte/components/SortableTable/SortableTable.YH4C4BSC.js',
        '/dev/_elderjs/props/ejs-983880793.js',
      ],
      [
        'headerNMYVZSgVzf',
        '/dev/_elderjs/svelte/components/Header/Header.AOWJN766.js',
        { permalink: '/dev/providers/', server: true },
      ],
    ]);
  },
  { timeout: 1000 },
);

```

**Reply by gregod** [reply]
I started some incremental work here https://github.com/Elderjs/elderjs/compare/v1.5...gregod:feature-minify-hydration-code but am unfortunately busy with some other things in the near future.  I will leave some comments on your commit.

**Reply by nickreese** [reply]
Shipped with `v1.5.0`.
